File: akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala
/**
 * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package docs.akka.typed

//#imports
import akka.typed._
import akka.typed.scaladsl.Actor._
import akka.typed.scaladsl.AskPattern._
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.concurrent.Await
//#imports
import akka.testkit.AkkaSpec
import akka.typed.TypedSpec

object IntroSpec {

  //#hello-world-actor
  object HelloWorld {
    final case class Greet(whom: String, replyTo: ActorRef[Greeted])
    final case class Greeted(whom: String)

    val greeter = Stateless[Greet] { (_, msg) ⇒
      println(s"Hello ${msg.whom}!")
      msg.replyTo ! Greeted(msg.whom)
    }
  }
  //#hello-world-actor

  //#chatroom-actor
  object ChatRoom {
    //#chatroom-protocol
    sealed trait Command
    final case class GetSession(screenName: String, replyTo: ActorRef[SessionEvent])
      extends Command
    //#chatroom-protocol
    //#chatroom-behavior
    private final case class PostSessionMessage(screenName: String, message: String)
      extends Command
    //#chatroom-behavior
    //#chatroom-protocol

    sealed trait SessionEvent
    final case class SessionGranted(handle: ActorRef[PostMessage]) extends SessionEvent
    final case class SessionDenied(reason: String) extends SessionEvent
    final case class MessagePosted(screenName: String, message: String) extends SessionEvent

    final case class PostMessage(message: String)
    //#chatroom-protocol
    //#chatroom-behavior

    def chatRoom(sessions: List[ActorRef[SessionEvent]] = List.empty): Behavior[Command] =
      Stateful[Command] { (ctx, msg) ⇒
        msg match {
          case GetSession(screenName, client) ⇒
            val wrapper = ctx.spawnAdapter {
              p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
            }
            client ! SessionGranted(wrapper)
            chatRoom(client :: sessions)
          case PostSessionMessage(screenName, message) ⇒
            val mp = MessagePosted(screenName, message)
            sessions foreach (_ ! mp)
            Same
        }
      }
    //#chatroom-behavior

    //#chatroom-actor
    def mutableChatRoom(): Behavior[Command] =
      Deferred[Command] { _ =>
        var sessions: List[ActorRef[SessionEvent]] = List.empty
        Stateful[Command] { (ctx, msg) ⇒
          msg match {
            case GetSession(screenName, client) ⇒
              val wrapper = ctx.spawnAdapter {
                p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
              }
              client ! SessionGranted(wrapper)
              sessions = client :: sessions
              Same
            case PostSessionMessage(screenName, message) ⇒
              val mp = MessagePosted(screenName, message)
              sessions foreach (_ ! mp)
              Same
          }
        }
      }
    //#chatroom-actor
  }
  //#chatroom-actor

}

class IntroSpec extends TypedSpec {
  import IntroSpec._

  def `must say hello`(): Unit = {
    // TODO Implicits.global is not something we would like to encourage in docs
    //#hello-world
    import HelloWorld._
    // using global pool since we want to run tasks after system.terminate
    import scala.concurrent.ExecutionContext.Implicits.global

    val system: ActorSystem[Greet] = ActorSystem("hello", greeter)

    val future: Future[Greeted] = system ? (Greet("world", _))

    for {
      greeting ← future.recover { case ex ⇒ ex.getMessage }
      done ← { println(s"result: $greeting"); system.terminate() }
    } println("system terminated")
    //#hello-world
  }

  def `must chat`(): Unit = {
    //#chatroom-gabbler
    import ChatRoom._

    val gabbler =
      Stateful[SessionEvent] { (_, msg) ⇒
        msg match {
          case SessionDenied(reason) ⇒
            println(s"cannot start chat room session: $reason")
            Stopped
          case SessionGranted(handle) ⇒
            handle ! PostMessage("Hello World!")
            Same
          case MessagePosted(screenName, message) ⇒
            println(s"message has been posted by '$screenName': $message")
            Stopped
        }
      }
    //#chatroom-gabbler

    //#chatroom-main
    val main: Behavior[akka.NotUsed] =
      Deferred { ctx =>
        val chatRoom = ctx.spawn(ChatRoom.chatRoom(), "chatroom")
        val gabblerRef = ctx.spawn(gabbler, "gabbler")
        ctx.watch(gabblerRef)
        chatRoom ! GetSession("ol’ Gabbler", gabblerRef)

        Stateful(
          onMessage = (_, _) ⇒ Unhandled,
          onSignal = (ctx, sig) ⇒
          sig match {
            case Terminated(ref) ⇒
              Stopped
            case _ ⇒
              Unhandled
          }
        )
      }

    val system = ActorSystem("ChatRoomDemo", main)
    Await.result(system.whenTerminated, 1.second)
    //#chatroom-main
  }

}

File: akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java
/**
 * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package jdocs.akka.typed;

//#imports
import akka.typed.ActorRef;
import akka.typed.Behavior;
import akka.typed.ExtensibleBehavior;
import akka.typed.Signal;
import akka.typed.javadsl.Actor;
import akka.typed.javadsl.ActorContext;
//#imports
import java.util.ArrayList;
import java.util.List;

public class IntroSpec {

  //#hello-world-actor
  public static class HelloWorld {
    public static final class Greet{
      public final String whom;
      public final ActorRef<Greeted> replyTo;

      public Greet(String whom, ActorRef<Greeted> replyTo) {
        this.whom = whom;
        this.replyTo = replyTo;
      }
    }

    public static final class Greeted {
      public final String whom;

      public Greeted(String whom) {
        this.whom = whom;
      }
    }

    public static final Behavior<Greet> greeter = Actor.stateless((ctx, msg) -> {
      System.out.println("Hello " + msg.whom + "!");
      msg.replyTo.tell(new Greeted(msg.whom));
    });
  }
  //#hello-world-actor

  //#chatroom-actor
  public static class ChatRoom {
    //#chatroom-protocol
    static interface Command {}
    public static final class GetSession implements Command {
      public final String screenName;
      public final ActorRef<SessionEvent> replyTo;
      public GetSession(String screenName, ActorRef<SessionEvent> replyTo) {
        this.screenName = screenName;
        this.replyTo = replyTo;
      }
    }
    //#chatroom-protocol
    //#chatroom-behavior
    private static final class PostSessionMessage implements Command {
      public final String screenName;
      public final String message;
      public PostSessionMessage(String screenName, String message) {
        this.screenName = screenName;
        this.message = message;
      }
    }
    //#chatroom-behavior
    //#chatroom-protocol

    static interface SessionEvent {}
    public static final class SessionGranted implements SessionEvent {
      public final ActorRef<PostMessage> handle;
      public SessionGranted(ActorRef<PostMessage> handle) {
        this.handle = handle;
      }
    }
    public static final class SessionDenied implements SessionEvent {
      public final String reason;
      public SessionDenied(String reason) {
        this.reason = reason;
      }
    }
    public static final class MessagePosted implements SessionEvent {
      public final String screenName;
      public final String message;
      public MessagePosted(String screenName, String message) {
        this.screenName = screenName;
        this.message = message;
      }
    }

    public static final class PostMessage {
      public final String message;
      public PostMessage(String message) {
        this.message = message;
      }
    }
    //#chatroom-protocol
    //#chatroom-behavior

    public static Behavior<Command> chatRoom() {
      return chatRoom(new ArrayList<ActorRef<SessionEvent>>());
    }

    private static Behavior<Command> chatRoom(List<ActorRef<SessionEvent>> sessions) {
      return Actor.stateful((ctx, msg) -> {
        if (msg instanceof GetSession) {
          GetSession getSession = (GetSession) msg;
          ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
            new PostSessionMessage(getSession.screenName, p.message));
          getSession.replyTo.tell(new SessionGranted(wrapper));
          // TODO mutable collection :(
          List<ActorRef<SessionEvent>> newSessions = new ArrayList<ActorRef<SessionEvent>>(sessions);
          newSessions.add(getSession.replyTo);
          return chatRoom(newSessions);
        } else if (msg instanceof PostSessionMessage) {
          PostSessionMessage post = (PostSessionMessage) msg;
          MessagePosted mp = new MessagePosted(post.screenName, post.message);
          sessions.forEach(s -> s.tell(mp));
          return Actor.same();
        } else {
          return Actor.unhandled();
        }
      });
    }
    //#chatroom-behavior

    public static Behavior<Command> mutableChatRoom() {
      return Actor.deferred(c -> {
        final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
        int postCount = 0;

        return Actor.stateful((ctx, msg) -> {
          if (msg instanceof GetSession) {
            GetSession getSession = (GetSession) msg;
            ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
              new PostSessionMessage(getSession.screenName, p.message));
            getSession.replyTo.tell(new SessionGranted(wrapper));
            sessions.add(getSession.replyTo);
            return Actor.same();
          } else if (msg instanceof PostSessionMessage) {
            PostSessionMessage post = (PostSessionMessage) msg;
            MessagePosted mp = new MessagePosted(post.screenName, post.message);
            // Local variable postCount defined in an enclosing scope must be final or effectively final
            //postCount += 1;
            sessions.forEach(s -> s.tell(mp));
            return Actor.same();
          } else {
            return Actor.unhandled();
          }
        });
      });
    }

    public static Behavior<Command> mutableChatRoom2() {
      // TODO Type safety: The expression of type IntroSpec.ChatRoom.MutableChatRoom2 needs unchecked conversion to
      //      conform to Behavior<IntroSpec.ChatRoom.Command>
      return Actor.deferred(ctx -> new MutableChatRoom2());
    }

    public static class MutableChatRoom2 extends ExtensibleBehavior<Command> {
      final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
      int postCount = 0;

      @Override
      public Behavior<Command> message(akka.typed.ActorContext<Command> ctx, Command msg) {
        if (msg instanceof GetSession) {
          GetSession getSession = (GetSession) msg;
          ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
            new PostSessionMessage(getSession.screenName, p.message));
          getSession.replyTo.tell(new SessionGranted(wrapper));
          sessions.add(getSession.replyTo);
          return Actor.same();
        } else if (msg instanceof PostSessionMessage) {
          PostSessionMessage post = (PostSessionMessage) msg;
          MessagePosted mp = new MessagePosted(post.screenName, post.message);
          postCount += 1;
          sessions.forEach(s -> s.tell(mp));
          return Actor.same();
        } else {
          return Actor.unhandled();
        }
      }

      @Override
      public Behavior<Command> management(akka.typed.ActorContext<Command> ctx, Signal msg) {
        return Actor.same();
      }
    }

    public static Behavior<Command> mutableChatRoom3() {
      return Actor.mutable(ctx -> new MutableChatRoom3(ctx));
    }

    public static class MutableChatRoom3 extends Actor.MutableBehavior<Command> {
      final ActorContext<Command> ctx;
      final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();

      public MutableChatRoom3(ActorContext<Command> ctx) {
        this.ctx = ctx;
      }

      @Override
      public Behavior<Command> onMessage(Command msg) {
        if (msg instanceof GetSession) {
          GetSession getSession = (GetSession) msg;
          ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
            new PostSessionMessage(getSession.screenName, p.message));
          getSession.replyTo.tell(new SessionGranted(wrapper));
          sessions.add(getSession.replyTo);
          return Actor.same();
        } else if (msg instanceof PostSessionMessage) {
          PostSessionMessage post = (PostSessionMessage) msg;
          MessagePosted mp = new MessagePosted(post.screenName, post.message);
          sessions.forEach(s -> s.tell(mp));
          return Actor.same();
        } else {
          return Actor.unhandled();
        }
      }

    }


  }
  //#chatroom-actor

}
